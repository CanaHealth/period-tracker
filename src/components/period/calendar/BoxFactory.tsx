import { useState } from 'react';

import clsxm from '@/lib/clsxm';

import Box from '@/components/period/calendar/Box';
import { FlowIntensity } from '@/components/period/calendar/options/NoteFlow';

const setColorVariant = (howHeavy: FlowIntensity) => {
  switch (howHeavy) {
    case 'heavy':
      return 'flowHeavy';
    case 'light':
      return 'flowLight';
    case 'average':
      return 'flow';
    case 'none':
      return 'normal';
    default:
      return undefined;
  }
};

function useLocalStorage(key: string, initialValue: string) {
  // State to store our value
  // Pass initial state function to useState so logic is only executed once
  const [storedValue, setStoredValue] = useState(() => {
    if (typeof window === 'undefined') {
      return initialValue;
    }
    try {
      // Get from local storage by key
      const item = window.localStorage.getItem(key);
      // Parse stored json or if none return initialValue
      return item ? JSON.parse(item) : initialValue;
    } catch (error) {
      // If error also return initialValue
      return initialValue;
    }
  });
  // Return a wrapped version of useState's setter function that ...
  // ... persists the new value to localStorage.
  const setValue = (value) => {
    try {
      // Allow value to be a function so we have same API as useState
      const valueToStore =
        value instanceof Function ? value(storedValue) : value;
      // Save state
      setStoredValue(valueToStore);
      // Save to local storage
      if (typeof window !== 'undefined') {
        window.localStorage.setItem(key, JSON.stringify(valueToStore));
      }
    } catch (error) {
      // A more advanced implementation would handle the error case
      console.log(error);
    }
  };
  return [storedValue, setValue];
}

const manageLocalStorage = (dateString: string, nextHowHeavy: string) => {
  if (nextHowHeavy === 'none') {
    localStorage.removeItem(dateString);
  } else {
    localStorage.setItem(dateString, nextHowHeavy);
  }
};

const howHeavyOptions: FlowIntensity[] = ['none', 'light', 'average', 'heavy'];

export type BoxFactoryInputs = {
  date: Date;
  color: FlowIntensity;
} & React.ComponentPropsWithoutRef<'div'>;

const BoxFactory: React.FC<BoxFactoryInputs> = ({ date, color }) => {
  const [open, setOpen] = useState(false);
  const [howHeavy, setHowHeavy] = useLocalStorage(
    String(date.getTime()),
    color
  );

  const handleDoubleClick = () => {
    setOpen(!open);
  };

  const handleClick = () => {
    const nextHowHeavy =
      howHeavyOptions[
        (howHeavyOptions.indexOf(howHeavy) + 1) % howHeavyOptions.length
      ];

    setHowHeavy(nextHowHeavy);

    const dateString = String(date.getTime());
    manageLocalStorage(dateString, nextHowHeavy);
  };

  const ColorVariant = setColorVariant(howHeavy);

  return (
    <div className={clsxm(' m-1 sm:m-2', '')}>
      {/* <NoteFlow
        flowdata={flowdata}
        open={open}
        setOpen={setOpen}
        handleSubmit={setFlowdata}
      /> */}

      <button
        data-testid='handle-flow-change'
        onClick={handleClick}
        onDoubleClick={handleDoubleClick}
      >
        <Box color={ColorVariant} date={date} />
      </button>
    </div>
  );
};

export default BoxFactory;

/*

const getLocalArray = (flowData: FlowData, localFlowData: string) => {
  if (localFlowData) {
    const localFlowDataArray = JSON.parse(localFlowData);
    const currentFlowData = flowData.date.getTime();

    const index = localFlowDataArray.findIndex((el: FlowData) => {
      const localDateHolder = new Date(el.date);
      return localDateHolder.getTime() === currentFlowData;
    });

    if (index !== -1) {
      return localFlowDataArray[index];
    }
  }
};

const manageLocalStorageArray = (nextFlowData: FlowData) => {
  const localFlowData = localStorage.getItem('FLOWDATA');

  if (localFlowData) {
    const localFlowDataArray = JSON.parse(localFlowData);
    const nextFlowDataDate = nextFlowData.date.getTime();

    const index = localFlowDataArray.findIndex((el: FlowData) => {
      const localDateHolder = new Date(el.date);
      return localDateHolder.getTime() === nextFlowDataDate;
    });

    if (index !== -1) {
      if (nextFlowData.howHeavy === 'none') {
        localFlowDataArray.splice(index, 1);
      } else {
        localFlowDataArray[index] = nextFlowData;
      }
    } else {
      localFlowDataArray.push(nextFlowData);
    }
    localStorage.setItem('FLOWDATA', JSON.stringify(localFlowDataArray));
  } else {
    localStorage.setItem('FLOWDATA', JSON.stringify([nextFlowData]));
  }
};

*/
