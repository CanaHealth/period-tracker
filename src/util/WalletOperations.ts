import { Keypair } from '@solana/web3.js';
import {
  createCipheriv,
  createDecipheriv,
  createHash,
  randomBytes,
} from 'crypto';

export type HumanReadableSolanaWallet = {
  publicKey: string;
  secretKey: string | Uint8Array;
  authTag?: string;
};

const ivSize = 16; // Initialization vector size in bytes.
const keySize = 32;

const createSolanaWallet: Keypair = Keypair.generate();

/**
 * `createSolanaWalletInHumanReadableFormat` creates a human readable format of a solana wallet.
 * @returns {HumanReadableSolanaWallet} A human readable format of a solana wallet.
 */
const createSolanaWalletInHumanReadableFormat = () => {
  const { publicKey, secretKey } = createSolanaWallet;
  return {
    publicKey: publicKey.toBase58(),
    secretKey: secretKey.toString(),
  };
};

/**
 * `saveWalletToLocalStorage` saves a human readable format of a solana wallet to local storage.
 * @param encryptedWallet The encrypted wallet to save.
 */
const storeWalletInLocalStorage = (
  encryptedWallet: HumanReadableSolanaWallet
) => {
  localStorage.setItem('solanaWallet', JSON.stringify(encryptedWallet));
};

/**
 * `checkForWalletInLocalStorage` checks if a wallet is stored in local storage.
 * @returns {boolean} True if a wallet is stored in local storage. False otherwise.
 */
const checkForWalletInLocalStorage = () => {
  if (localStorage.getItem('solanaWallet') === null) {
    return false;
  } else {
    return true;
  }
};

/**
 * `getWalletFromLocalStorage` gets a human readable format of a solana wallet from local storage.
 * @returns {HumanReadableSolanaWallet} A human readable format of a solana wallet from the local storage.
 */
const getWalletFromLocalStorage = () => {
  return JSON.parse(localStorage.getItem('solanaWallet')!);
};

/**
 * `hashPasscode` hashes a passcode with the sha256 hashing algorithm.
 * @param passcode The passcode to hash.
 * @returns {string} The hashed passcode.
 */
const hashPasscode = (passcode: string) => {
  return createHash('sha256').update(passcode).digest('hex');
};

/**
 * `encryptWallet` encrypts the secret key in human readable format of a solana wallet.
 * @param wallet The unencrypted human readable format of a solana wallet.
 * @param passcode The passcode to encrypt the secret key with.
 * @returns {HumanReadableSolanaWallet} Solana wallet with encrypted secret key.
 */
const encryptWallet = (wallet: HumanReadableSolanaWallet, passcode: string) => {
  const hashedPasscode: string = hashPasscode(passcode).slice(0, keySize); // Get first 32 characters of the hashed passcode.
  const iv: string = randomBytes(ivSize).toString('hex').slice(0, ivSize); // iv is 16 bytes long
  // Encrypt secretKey with AES-256-GCM using hashedpasscode as key and random bytes as iv.
  const cipher = createCipheriv(
    'aes-256-gcm', // algorithm
    hashedPasscode, // key
    iv // iv
  );
  const encryptedSecretKey =
    cipher.update(wallet.secretKey, 'utf-8', 'hex') + cipher.final('hex');

  const encryptedSecretKeyWithIV: string = encryptedSecretKey + iv;

  return {
    publicKey: wallet.publicKey,
    secretKey: encryptedSecretKeyWithIV,
    authTag: cipher.getAuthTag().toString('hex'),
  };
};

/**
 * `decryptWallet` decrypts the secret key in human readable format of a solana wallet.
 * @param wallet The human readable format of a solana wallet with an encrypted secret key.
 * @param passcode The passcode to decrypt the secret key with.
 * @returns {HumanReadableSolanaWallet} Solana wallet with decrypted secret key.
 */
const decryptWallet = (wallet: HumanReadableSolanaWallet, passcode: string) => {
  const hashedPasscode: string = hashPasscode(passcode).slice(0, keySize); // Get first 32 characters of the hashed passcode.
  const encryptedSecretKey = wallet.secretKey;
  const iv = encryptedSecretKey.slice(encryptedSecretKey.length - ivSize); // Get last 16 characters of the initialization vector appended to encrypted data.
  const encryptedSecretKeyWithoutIV = encryptedSecretKey.slice(
    0,
    encryptedSecretKey.length - ivSize
  ); // Remove last 16 characters of the initialization vector appended to encrypted data.
  const decipher = createDecipheriv(
    'aes-256-gcm', // algorithm
    hashedPasscode, // key
    iv // iv
  );

  decipher.setAuthTag(Buffer.from(wallet.authTag, 'hex')); // Set the authentication tag.

  try {
    const decryptedSecretKey =
      decipher.update(encryptedSecretKeyWithoutIV, 'hex', 'utf-8') +
      decipher.final('utf-8');

    return {
      publicKey: wallet.publicKey,
      secretKey: converSKToArray(decryptedSecretKey),
    };
  } catch (error) {
    console.log(error);
    return {
      publicKey: wallet.publicKey,
      secretKey: '',
    };
  }
};

/**
 * `convertSKToArray` converts a secret key to an array.
 * @param {string} secretKey The secret key to convert.
 * @returns {Uint8Array} The secret key as a Uint8Array - as definded by the @solana/web3.js library.
 */
const converSKToArray = (secretKey: string) => {
  return Uint8Array.from(secretKey.split(',').map((item) => parseInt(item)));
};

export {
  checkForWalletInLocalStorage,
  createSolanaWalletInHumanReadableFormat,
  decryptWallet,
  encryptWallet,
  getWalletFromLocalStorage,
  hashPasscode,
  storeWalletInLocalStorage,
};
