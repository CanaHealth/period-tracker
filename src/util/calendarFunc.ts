import { FlowData } from '@/components/period/calendar/options/NoteFlow';

const normalizeDate = (date: Date): Date => {
  const newDate = new Date(date);

  newDate.setHours(8);
  newDate.setMinutes(0);
  newDate.setSeconds(0);
  newDate.setMilliseconds(0);

  return newDate;
};

const today = normalizeDate(new Date());

const newToday = () => {
  return today;
};

const dayOfMiliseconds = 1000 * 60 * 60 * 24;
const weekOfMiliseconds = dayOfMiliseconds * 7;

const daysFrom = (date: Date, days: number) => {
  return normalizeDate(new Date(date.getTime() + dayOfMiliseconds * days));
};

const weeksFrom = (date: Date, weeks: number) => {
  return normalizeDate(new Date(date.getTime() + weekOfMiliseconds * weeks));
};

const findPrevMonday = (date: Date) => {
  const monday = new Date(date);
  monday.setDate(date.getDate() - ((date.getDay() + 6) % 7));
  return normalizeDate(monday);
};

const weekOfDates = (monday: Date, flowInfoFromStorage: FlowData): FlowData => {
  const days: FlowData = {};
  const mondayInTest = findPrevMonday(monday);
  const defaultFlow = 'none';

  for (let i = 0; i < 7; i++) {
    const step = i;
    const day = daysFrom(mondayInTest, step);
    const timestamp = day.getTime();
    const dayFlow = flowInfoFromStorage[timestamp] || defaultFlow;

    days[timestamp] = dayFlow;
  }

  return days;
};

const manyWeeks = (
  numWeeks: number,
  flowInfoFromStorage: FlowData
): FlowData => {
  const weeks = {};
  for (let i = 0; i < numWeeks; i++) {
    const step = i;
    const monday = weeksFrom(findPrevMonday(today), step);
    const week = weekOfDates(monday, flowInfoFromStorage);
    Object.assign(weeks, week);
  }
  return weeks;
};

export {
  dayOfMiliseconds,
  daysFrom,
  findPrevMonday,
  manyWeeks,
  newToday,
  normalizeDate,
  weekOfDates,
  weekOfMiliseconds,
  weeksFrom,
};

/*
const weekFactory = (date: Date, num: number) => {
  const weeks: { days: Date[] }[] = [];
  for (let i = 0; i < num; i++) {
    const step = i * 7;
    const weekToProcess = new Date(date.getTime());
    weekToProcess.setDate(date.getDate() - step);
    const mondayOfWeek = findMonday(weekToProcess);
    const days: Date[] = [];

    for (let j = 0; j < 7; j++) {
      const step = j;

      const day = new Date(mondayOfWeek.getTime());
      day.setDate(mondayOfWeek.getDate() + step);

      const howHeavy = 'none';

      days.push({ howHeavy, date: day });
    }
    weeks.unshift({ days });
  }

  return weeks;
};
const weeks = weekFactory(today, 5);

*/
