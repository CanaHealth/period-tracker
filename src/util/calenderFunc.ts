const normalizeDate = (date: Date): Date => {
  const newDate = new Date(date);

  newDate.setHours(8);
  newDate.setMinutes(0);
  newDate.setSeconds(0);
  newDate.setMilliseconds(0);

  return newDate;
};

const today = normalizeDate(new Date());

const newToday = () => {
  return today;
};

const dayOfMiliseconds = 1000 * 60 * 60 * 24;
const weekOfMiliseconds = dayOfMiliseconds * 7;

const daysFrom = (date: Date, days: number) => {
  return normalizeDate(new Date(date.getTime() + dayOfMiliseconds * days));
};

const weeksFrom = (date: Date, weeks: number) => {
  return normalizeDate(new Date(date.getTime() + weekOfMiliseconds * weeks));
};

const findPrevMonday = (date: Date) => {
  const monday = new Date(date);
  monday.setDate(date.getDate() - ((date.getDay() + 6) % 7));
  return normalizeDate(monday);
};

const mockArrayOfFlowData = [
  {
    '1660550400000': 'light',
  },
  {
    '1660636800000': 'average',
  },
  {
    '1660723200000': 'heavy',
  },
  {
    '1660896000000': 'heavy',
  },
  {
    '1661068800000': 'heavy',
  },
];

const weekOfDates = (monday: Date) => {
  const days = [];
  const mondayInTest = findPrevMonday(monday);
  const defaultFlow = 'none';

  for (let i = 0; i < 7; i++) {
    const step = i;
    const day = new Date(mondayInTest.getTime());
    day.setDate(mondayInTest.getDate() + step);
    const timestamp = daysFrom(day, step).getTime();
    const dayFlow = mockArrayOfFlowData.find((day) => day[timestamp]);

    if (dayFlow) {
      days.push(dayFlow);
    } else {
      days.push({ [timestamp]: defaultFlow });
    }
  }

  return days;
};

const manyWeeks = (monday: Date, weeks: number) => {
  const weeksArray = [];
  for (let i = 0; i < weeks; i++) {
    const step = i;
    const monday = weeksFrom(today, step);
    weeksArray.push(weekOfDates(monday));
  }
  return weeksArray;
};

export {
  dayOfMiliseconds,
  daysFrom,
  findPrevMonday,
  manyWeeks,
  newToday,
  normalizeDate,
  weekOfDates,
  weekOfMiliseconds,
  weeksFrom,
};

/*
const weekFactory = (date: Date, num: number) => {
  const weeks: { days: Date[] }[] = [];
  for (let i = 0; i < num; i++) {
    const step = i * 7;
    const weekToProcess = new Date(date.getTime());
    weekToProcess.setDate(date.getDate() - step);
    const mondayOfWeek = findMonday(weekToProcess);
    const days: Date[] = [];

    for (let j = 0; j < 7; j++) {
      const step = j;

      const day = new Date(mondayOfWeek.getTime());
      day.setDate(mondayOfWeek.getDate() + step);

      const howHeavy = 'none';

      days.push({ howHeavy, date: day });
    }
    weeks.unshift({ days });
  }

  return weeks;
};
const weeks = weekFactory(today, 5);

*/
