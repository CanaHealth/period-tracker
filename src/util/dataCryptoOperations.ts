/**
 * `dataCryptoOperations.ts` is a collection of functions that are used to encrypt and decrypt data using the AES-256 encryption algorithm.
 */
import {encode } from "@stablelib/base64";
import {
  createCipheriv,
  createDecipheriv,
  randomBytes
} from 'crypto';


const ivSize = 16; // Initialization vector size in bytes.
const keySize = 32;

/**
 * `AESGCMCipherText` is a type that represents the cipherText and authTag. Used for AES-256-GCM encryption.
 */
export type AESGCMCipherText = {
  cipherText: string;
  authTag: string;
}

/**
 * `encryptData` is used to encrypt data using the AES-256 encryption algorithm.
 * @param {string} data The data to encrypt.
 * @param {Uint8Array} secretKey Solana wallet secret key in Uint8Array format.
 * @returns {AESGCMCipherText} cipherText with authTag.
 */
const encryptData = (
  data: string,
  secretKey: Uint8Array
): AESGCMCipherText => {
  const iv: string = randomBytes(ivSize).toString('hex').slice(0, ivSize); // iv is 16 bytes long
  const encodedSecretKey: string = encode(secretKey).slice(0, keySize);
  // Encrypt secretKey with AES-256-GCM using hashedPasscode  as key and random bytes as iv.
  const cipher = createCipheriv(
    'aes-256-gcm', // algorithm
    encodedSecretKey, // key
    iv // iv
  );

  const encryptedSecretKey =
    cipher.update(data, 'utf-8', 'hex') +
    cipher.final('hex');

  const encryptedSecretKeyWithIV: string = encryptedSecretKey + iv;

  return {
    cipherText: encryptedSecretKeyWithIV,
    authTag: cipher.getAuthTag().toString('hex'),
  };
}

/**
 * `decryptData` is used to decrypt cipherText using the AES-256 encryption algorithm.
 * @param AESGCMCipherText cipherText with authTag. 
 * @param {Uint8Array} secretKey decrypted Solana wallet secret key in Uint8Array format.
 * @returns 
 */
const decryptData = (
  { cipherText, authTag }: AESGCMCipherText,
  secretKey: Uint8Array
): string => {
  if (authTag === undefined) {
    throw new Error('AuthTag is undefined or secretKey cannot be a decrypted.');
  } else {
    const encodedSecretKey: string = encode(secretKey).slice(0, keySize);
    const iv = cipherText.slice(cipherText.length - ivSize); // Get last 16 characters of the initialization vector appended to encrypted data.

    // encrypted secret key should be a
    const encryptedSecretKeyWithoutIV = cipherText.slice(
      0,
      cipherText.length - ivSize
    ); // Remove last 16 characters of the initialization vector appended to encrypted data.
    const decipher = createDecipheriv(
      'aes-256-gcm', // algorithm
      encodedSecretKey, // key
      iv // iv
    );

    decipher.setAuthTag(Buffer.from(authTag, 'hex')); // Set the authentication tag.

    try {
      const decreyptedData =
        decipher.update(encryptedSecretKeyWithoutIV, 'hex', 'utf-8') +
        decipher.final('utf-8');

      return decreyptedData;
    } catch (error) {
      throw new Error('Invalid key entered');
    }
  }
};

export { decryptData,encryptData};