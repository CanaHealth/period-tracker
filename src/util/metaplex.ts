/**
 * `metaplex.ts` initializes and uses the MetaPlex JS SDK.
 */
import { keypairIdentity, Metaplex } from "@metaplex-foundation/js";
import { clusterApiUrl, Connection, Keypair } from "@solana/web3.js";
import { NFTStorageMetaplexor } from '@nftstorage/metaplex-auth'

function initializeMetaplex() {
  const connection = new Connection(clusterApiUrl("devnet"));
  
  const metaplex = Metaplex.make(connection);

  return metaplex;
}

async function createNFT(data: Object, secretKey: Uint8Array) {
  const wallet = Keypair.fromSecretKey(secretKey);

  const metaplex = initializeMetaplex();
  metaplex.use(keypairIdentity(wallet));
  
  const cid = await uploadNFT(data, secretKey);
  console.log("CID: ", cid);

  const { nft } = await metaplex
  .nfts()
  .create({
      uri: "ipfs://" + cid,
      name: "Cana",
      sellerFeeBasisPoints: 500, // Represents 5.00%.
  })
  .run();

  console.log(nft);
  console.log(nft.mint.address.toString());

  return nft
}

async function uploadNFT(data: Object, key: Uint8Array) {
  const client = NFTStorageMetaplexor.withSecretKey(key, {
    solanaCluster: 'devnet',
    mintingAgent: 'canahealth/period_tracker'
  });

  const blob: Blob = new Blob([JSON.stringify(data)]);
  const cid = await client.storeBlob(blob);

  return cid;
}

export { createNFT,initializeMetaplex };