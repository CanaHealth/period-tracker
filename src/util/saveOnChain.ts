import { encryptData } from './dataCryptoOperations'
import { createNFT } from './metaplex'

const saveDataOnChain = async (data: any, secretKey: Uint8Array) => {
  // Encrypt data with secretKey
  const encryptedData = encryptData(data, secretKey)
  // Mint the encrypted data on chain.
  const nftTokenID = await createNFT(encryptedData, secretKey)
  // store nft token id in local storage (temporarily).
  localStorage.setItem('nftTokenID', nftTokenID)

  return nftTokenID
}

export { saveDataOnChain }
