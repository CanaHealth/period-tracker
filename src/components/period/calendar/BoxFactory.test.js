/* eslint-disable unused-imports/no-unused-imports */
import { afterAll, clearAllMocks, describe, expect, it } from '@jest/globals';
import { fireEvent, render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { newToday } from '@/util/calendarFunc';

import BoxFactory from './BoxFactory';

// -----------------------------------------------------------------------------
// test box factory local storage behavoir
// -----------------------------------------------------------------------------

const unixTimeStamp = 1660550400000;

const date = new Date(unixTimeStamp);

const defaultTestID = 'handle-flow-change';

describe('BoxFactory', () => {
  it('renders without crashing', () => {
    render(<BoxFactory date={date} />);
  });

  // snapshot
  it('matches snapshot', () => {
    const BoxComp = render(<BoxFactory date={date} />);
    expect(BoxComp.asFragment()).toMatchSnapshot();
  });

  it('has a11y violations', async () => {
    const BoxComp = render(<BoxFactory date={date} />);
    const results = await axe(BoxComp.container);
    expect(results).toHaveNoViolations();
  });

  it('got 1 click - saved howheavy to localStorage', () => {
    const BoxComp = render(<BoxFactory date={date} />);
    const button = BoxComp.getByTestId(defaultTestID);
    const dateString = String(unixTimeStamp);
    fireEvent.click(button);
    expect(localStorage.getItem(dateString)).toEqual('light');
  });

  it('got 2 clicks - saved howheavy to localStorage', () => {
    const BoxComp = render(<BoxFactory date={date} />);
    const button = BoxComp.getByTestId(defaultTestID);
    const dateString = String(unixTimeStamp);
    fireEvent.click(button);
    fireEvent.click(button);
    expect(localStorage.getItem(dateString)).toEqual('average');
  });

  it('got 3 clicks - saved howheavy to localStorage', () => {
    const BoxComp = render(<BoxFactory date={date} />);
    const button = BoxComp.getByTestId(defaultTestID);
    const dateString = String(unixTimeStamp);
    fireEvent.click(button);
    fireEvent.click(button);
    fireEvent.click(button);
    expect(localStorage.getItem(dateString)).toEqual('heavy');
  });

  it('got 4 clicks - did not save howheavy to localStorage because we done save none ', () => {
    const BoxComp = render(<BoxFactory date={date} />);
    const button = BoxComp.getByTestId(defaultTestID);
    const dateString = String(unixTimeStamp);
    fireEvent.click(button);
    fireEvent.click(button);
    fireEvent.click(button);
    fireEvent.click(button);
    expect(localStorage.getItem(dateString)).toBeNull();
  });

  it('got 5 clicks - saved howheavy to localStorage', () => {
    const BoxComp = render(<BoxFactory date={date} />);
    const button = BoxComp.getByTestId(defaultTestID);
    const dateString = String(unixTimeStamp);
    fireEvent.click(button);
    fireEvent.click(button);
    fireEvent.click(button);
    fireEvent.click(button);
    fireEvent.click(button);
    expect(localStorage.getItem(dateString)).toEqual('light');
  });

  afterAll(() => {
    localStorage.clear();
  });
});
