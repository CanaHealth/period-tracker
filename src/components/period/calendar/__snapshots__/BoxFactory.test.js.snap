// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoxFactory matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex snap-end items-center justify-center"
  >
    <button
      data-testid="handle-flow-change"
    >
      <div
        class="group relative flex h-8 w-8 items-center justify-center rounded-md text-xs hover:border-0 hover:bg-gray-dark-dark hover:text-white border-gray-light border-2 bg-gray-mid-light text-gray-700 transition-transform duration-200 ease-in-out hover:scale-110 hover:shadow-md active:scale-100"
      >
        <div
          aria-hidden="false"
          class="absolute inset-0 flex items-center justify-center rounded-md text-xs border-gray-light bg-gray-mid-light border-0"
        >
          Mo
        </div>
        <span
          aria-hidden="true"
          class="opacity-0 group-hover:opacity-100 group-active:opacity-100 group-focus:opacity-100 hidden"
        >
          15
        </span>
      </div>
    </button>
  </div>
</DocumentFragment>
`;
